Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py

```diff
@@ -37,70 +39,45 @@ def validateInputs(self):
-        onlyfiles = [f for f in os.listdir(source_folder)
-                     if os.path.isfile(os.path.join(source_folder, f))]
-        historyfiles = [x for x in onlyfiles if x.endswith('.py')]
-
-        self.log().debug("Found {0} history files".format(len(historyfiles)))
-
-        # Read each history in as a list of tuples
-        original_lines = set()
-        for infile in historyfiles:
-            with open(os.path.join(source_folder, infile), 'r') as f:
-                for line in f:
-                    original_lines.add(line)
-
-        commands = []
-        # Remove any comment lines
-        for line in original_lines:
-            stripped_line = line.lstrip()
-            if stripped_line[0] is not '#':
-                commands.append(stripped_line)
-
-        self.log().debug("Found {0} unique commands".format(len(commands)))
-
-        # Add lists of histories together
-        all_unique_commands = [command.strip().split('#') for command in commands]
-
-        # Convert the datetime into a sortable integer
-        all_unique_commands = [(i[0], self._concatenate_iso_datetime(i[1]))
-                               for i in all_unique_commands]
-
-        # Sort the new list on datetime integer
-        all_unique_commands.sort(key=lambda time: (time[1]))
+        all_lines = []
+
+        #for fn in historyfiles:
+        for fn in glob.iglob(os.path.join(self.getPropertyValue(_recovery_folder), '*.py')):
+            with open(fn) as f:
+                tokens = tokenize.generate_tokens(f.readline)
+
+                line = None
+                for t in tokens:
+                    #Start a new line when we see a name
+                    if line is None and t[0] == tokenize.NAME:
+                        line = [t]
+                    # End the line when we see a logical line ending
+                    elif t[0] == tokenize.NEWLINE:
+                        # Only care about the line if it has a comment
+                        have_comment = any(x[0] == tokenize.COMMENT for x in line)
+                        if have_comment:
+                            all_lines.append(line)
+                        line = []
+                    # Everything in between we care about
+                    elif line is not None:
+                        line.append(t)
+
+        # l[-1][1][1:] is the comment string, with the preceeding hash stripped off
+        all_commands = [(''.join(t[1] for t in l[:-1]), l[-1][1][1:]) for l in all_lines]
+        # Remove duplicate commands by casting commands as a set
+        unique_commands = list(set(all_commands))
+
+        unique_commands.sort(key=lambda time: (time[1]))
 
         destination = self.getPropertyValue(_destination_file)
 
         self.log().debug("Writing commands to file")
         # Write to file
         with open(destination, 'w') as outfile:
-            for x in all_unique_commands:
+            for x in unique_commands:
```

Comment:
Is this just writing the lines back out as they were?

In that case can't we just skip rebuilding the line from the tokens above and just output the full line from the comment token?

Reply:
@DanNixon I'm not sure I follow. The list `unique_commands` has been stripped of replication and sorted on the timestamp. I can't see how we would get the correct order from the comment token.

Reply:
`''.join(t[1] for t in l[:-1])` above is recreating the line originally read from the file minus the comment.

`outfile.write('{} # {} \n'.format(x[0], x[1]))` is formatting a string containing the reconstructed string previously mentioned, a `#` and the comment minus the `#`. The result of this is the same as the full source line that is also stored in the comment token (save the newline).

So really we only need to store the comment tokens, sort them on the comment text and output the full source line (last element of the token tuple) here.

Reply:
I believe the idea behind it, is that we are removing duplicate commands from the list so that project recovery doesn't run into issues further down the road. With the method you are suggesting it does not seem to take this into account @DanNixon

Reply:
No problem - hard to describe these concepts in this forum. If you can drop by tomorrow that'd be great. I'll be out from about 9:25 - 10:10. Otherwise I'm about.